nodes/student_localiser.py

```diff
@@ -0,0 +1,660 @@
+		if self.resize[1] is not None:
+			global IMAGE_WIDTH
+			IMAGE_WIDTH = self.resize[1]
+		if self.resize[0] is None and self.resize[1] is None:
+			self.resize = None
+		self.last_image = None
+		self.patch_size = image_processing.parse_patch_size_parameter(rospy.get_param('/patch_size', (9,9)))
+		global CORR_SUB_SAMPLING, FIELD_OF_VIEW_DEG
+		self.image_recognition_threshold = rospy.get_param('/image_recognition_threshold', 0.1)
+		CORR_SUB_SAMPLING = rospy.get_param('/image_subsampling', 1)
+		FIELD_OF_VIEW_DEG = rospy.get_param('/image_field_of_view_width_deg', 2*60.6 + 2*27.0)
+		FIELD_OF_VIEW_RAD = math.radians(FIELD_OF_VIEW_DEG)
+		self.search_range = rospy.get_param('~search-range', 1)
+
+		# data saving
+		self.save_dir = os.path.expanduser(rospy.get_param('/data_save_dir', '~/miro/data/follow-straight_tests/5'))
+		self.save_full_res_images = rospy.get_param('/save_full_res_images', True)
+		self.save_full_res_images_at_goal = rospy.get_param('/save_full_res_images_at_goal', True)
+		self.last_full_res_image = None
+		self.save_gt_data = rospy.get_param('/save_gt_data', False)
+		self.publish_gt_goals = rospy.get_param('/publish_gt_goals', True)
+		if self.save_dir[-1] != '/':
+			self.save_dir += '/'
+		if not os.path.isdir(self.save_dir):
+			os.makedirs(self.save_dir)
+		if not os.path.isdir(self.save_dir+'norm/'):
+			os.makedirs(self.save_dir+'norm/')
+		if not os.path.isdir(self.save_dir+'pose/'):
+			os.makedirs(self.save_dir+'pose/')
+		if not os.path.isdir(self.save_dir+'offset/'):
+			os.makedirs(self.save_dir+'offset/')
+		if not os.path.isdir(self.save_dir+'correction/'):
+			os.makedirs(self.save_dir+'correction/')
+		if self.save_full_res_images or self.save_full_res_images_at_goal:
+			if not os.path.isdir(self.save_dir+'full/'):
+				os.makedirs(self.save_dir+'full/')
+		self.goal_number = 0
+
+		self.save_params()
+
+
+	def setup_publishers(self):
+		self.goal_pub = rospy.Publisher('goal', Goal, queue_size=1)
+		if self.publish_gt_goals:
+			self.goal_pose_pub = rospy.Publisher('goal_pose', PoseStamped, queue_size=1)
+		rospy.wait_for_service('match_image')
+		self.match_image = rospy.ServiceProxy('match_image', ImageMatch, persistent=True)
+
+	def setup_subscribers(self):
+		if not self.ready:
+			self.srv_ready = rospy.Service('ready_localiser', Trigger, self.on_ready)
+		self.sub_odom = rospy.Subscriber("odom", Odometry, self.process_odom_data, queue_size=1)
+		self.sub_images = rospy.Subscriber('image', Image, self.process_image_data, queue_size=1, buff_size=2**22)
+		self.tfBuffer = tf2_ros.Buffer()
+		self.tfListener = tf2_ros.TransformListener(self.tfBuffer)
+
+		# For Multi-Robot System
+		self.sub_teacher_pose = rospy.Subscriber("teacher_pose", Pose, self.append_teacher_pose, queue_size=1)
+
+	def append_teacher_pose(self, msg):
+		'''Append new teacher pose to the student list when ever it arrived'''
+		self.poses.append(msg)
```

Comment:
Yes that is true

Reply:
To prevent that, we can append those data on the teacher side.
And create a service, where student can send a pose/image index to request images.
(Still need better thought to the logic)